# KNN with scikit-learn - Lab

## Introduction

In this lab, you'll learn how to use scikit-learn's implementation of a KNN classifier on the classic Titanic dataset from Kaggle!
 

## Objectives

In this lab you will:

- Conduct a parameter search to find the optimal value for K 
- Use a KNN classifier to generate predictions on a real-world dataset 
- Evaluate the performance of a KNN model  


## Getting Started

Start by importing the dataset, stored in the `titanic.csv` file, and previewing it.

In [1]:
# Your code here
# Import pandas and set the standard alias 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline


# Import the data from 'titanic.csv' and store it in a pandas DataFrame 
raw_df = pd.read_csv("./titanic.csv")

# Print the head of the DataFrame to ensure everything loaded correctly 
raw_df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Great!  Next, you'll perform some preprocessing steps such as removing unnecessary columns and normalizing features.

## Preprocessing the data

Preprocessing is an essential component in any data science pipeline. It's not always the most glamorous task as might be an engaging data visual or impressive neural network, but cleaning and normalizing raw datasets is very essential to produce useful and insightful datasets that form the backbone of all data powered projects. This can include changing column types, as in: 


```python
df['col_name'] = df['col_name'].astype('int')
```
Or extracting subsets of information, such as: 

```python
import re
df['street'] = df['address'].map(lambda x: re.findall('(.*)?\n', x)[0])
```

> **Note:** While outside the scope of this particular lesson, **regular expressions** (mentioned above) are powerful tools for pattern matching! See the [regular expressions official documentation here](https://docs.python.org/3.6/library/re.html). 

Since you've done this before, you should be able to do this quite well yourself without much hand holding by now. In the cells below, complete the following steps:

1. Remove unnecessary columns (`'PassengerId'`, `'Name'`, `'Ticket'`, and `'Cabin'`) 
2. Convert `'Sex'` to a binary encoding, where female is `0` and male is `1` 
3. Detect and deal with any missing values in the dataset:  
    * For `'Age'`, replace missing values with the median age for the dataset  
    * For `'Embarked'`, drop the rows that contain missing values
4. One-hot encode categorical columns such as `'Embarked'` 
5. Store the target column, `'Survived'`, in a separate variable and remove it from the DataFrame  

In [2]:
raw_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [3]:
# Drop the unnecessary columns
df = raw_df.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1)  # Drop columns = axis =1
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [15]:
# Convert Sex to binary encoding
df['Sex'] = pd.get_dummies(df['Sex'], drop_first=True)
df.head(25)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0          0       3    1  22.0      1      0   7.2500        S
1          1       1    0  38.0      1      0  71.2833        C
2          1       3    0  26.0      0      0   7.9250        S
3          1       1    0  35.0      1      0  53.1000        S
4          0       3    1  35.0      0      0   8.0500        S
5          0       3    1  28.0      0      0   8.4583        Q
6          0       1    1  54.0      0      0  51.8625        S
7          0       3    1   2.0      3      1  21.0750        S
8          1       3    0  27.0      0      2  11.1333        S
9          1       2    0  14.0      1      0  30.0708        C
10         1       3    0   4.0      1      1  16.7000        S
11         1       1    0  58.0      0      0  26.5500        S
12         0       3    1  20.0      0      0   8.0500        S
13         0       3    1  39.0      1      5  31.2750        S
14         0       3    0  14.0      0      0   7.8542        S
15         1       2    0  55.0      0      0  16.0000        S
16         0       3    1   2.0      4      1  29.1250        Q
17         1       2    1  28.0      0      0  13.0000        S
18         0       3    0  31.0      1      0  18.0000        S
19         1       3    0  28.0      0      0   7.2250        C
20         0       2    1  35.0      0      0  26.0000        S
21         1       2    1  34.0      0      0  13.0000        S
22         1       3    0  15.0      0      0   8.0292        Q
23         1       1    1  28.0      0      0  35.5000        S
24         0       3    0   8.0      3      1  21.0750        S

In [16]:
# Find the number of missing values in each column
df.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [6]:
df.shape

(891, 8)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null uint8
Age         714 non-null float64
SibSp       891 non-null int64
Parch       891 non-null int64
Fare        891 non-null float64
Embarked    889 non-null object
dtypes: float64(2), int64(4), object(1), uint8(1)
memory usage: 49.7+ KB


In [17]:
df.describe()

Survived      Pclass         Sex         Age       SibSp       Parch  \
count  889.000000  889.000000  889.000000  889.000000  889.000000  889.000000   
mean     0.382452    2.311586    0.649044   29.315152    0.524184    0.382452   
std      0.486260    0.834700    0.477538   12.984932    1.103705    0.806761   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000    0.000000   22.000000    0.000000    0.000000   
50%      0.000000    3.000000    1.000000   28.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   35.000000    1.000000    0.000000   
max      1.000000    3.000000    1.000000   80.000000    8.000000    6.000000   

             Fare  
count  889.000000  
mean    32.096681  
std     49.697504  
min      0.000000  
25%      7.895800  
50%     14.454200  
75%     31.000000  
max    512.329200

In [9]:
# Impute the missing values in 'Age'
df['Age'] = df['Age'].fillna(df['Age'].median())
df.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [10]:
# Drop the rows missing values in the 'Embarked' column
df = df.dropna()
df.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [13]:
df.shape

(889, 8)

In [19]:
# One-hot encode the categorical columns
one_hot_df = pd.get_dummies(df)
one_hot_df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    1  22.0      1      0   7.2500           0           0   
1         1       1    0  38.0      1      0  71.2833           1           0   
2         1       3    0  26.0      0      0   7.9250           0           0   
3         1       1    0  35.0      1      0  53.1000           0           0   
4         0       3    1  35.0      0      0   8.0500           0           0   

   Embarked_S  
0           1  
1           0  
2           1  
3           1  
4           1

In [ ]:
# Assign the 'Survived' column to labels
labels = None

# Drop the 'Survived' column from one_hot_df


In [20]:
labels = one_hot_df['Survived']
one_hot_df.drop('Survived', axis=1, inplace=True)

In [22]:
one_hot_df.head(20)

Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0        3    1  22.0      1      0   7.2500           0           0   
1        1    0  38.0      1      0  71.2833           1           0   
2        3    0  26.0      0      0   7.9250           0           0   
3        1    0  35.0      1      0  53.1000           0           0   
4        3    1  35.0      0      0   8.0500           0           0   
5        3    1  28.0      0      0   8.4583           0           1   
6        1    1  54.0      0      0  51.8625           0           0   
7        3    1   2.0      3      1  21.0750           0           0   
8        3    0  27.0      0      2  11.1333           0           0   
9        2    0  14.0      1      0  30.0708           1           0   
10       3    0   4.0      1      1  16.7000           0           0   
11       1    0  58.0      0      0  26.5500           0           0   
12       3    1  20.0      0      0   8.0500           0           0   
13       3    1  39.0      1      5  31.2750           0           0   
14       3    0  14.0      0      0   7.8542           0           0   
15       2    0  55.0      0      0  16.0000           0           0   
16       3    1   2.0      4      1  29.1250           0           1   
17       2    1  28.0      0      0  13.0000           0           0   
18       3    0  31.0      1      0  18.0000           0           0   
19       3    0  28.0      0      0   7.2250           1           0   

    Embarked_S  
0            1  
1            0  
2            1  
3            1  
4            1  
5            0  
6            1  
7            1  
8            1  
9            0  
10           1  
11           1  
12           1  
13           1  
14           1  
15           1  
16           0  
17           1  
18           1  
19           0

## Create training and test sets

Now that you've preprocessed the data, it's time to split it into training and test sets. 

In the cell below:

* Import `train_test_split` from the `sklearn.model_selection` module 
* Use `train_test_split()` to split the data into training and test sets, with a `test_size` of `0.25`. Set the `random_state` to 42 

In [31]:
# Import train_test_split 
from sklearn.model_selection import train_test_split

# Split the data
X_train, X_test, y_train, y_test = train_test_split(one_hot_df, labels, test_size=0.25, random_state=42)
X_train.size, X_test.size, y_train.size, y_test.size


(5994, 2007, 666, 223)

## Normalizing the data

The final step in your preprocessing efforts for this lab is to **_normalize_** the data. We normalize **after** splitting our data into training and test sets. This is to avoid information "leaking" from our test set into our training set (read more about data leakage [here](https://machinelearningmastery.com/data-leakage-machine-learning/) ). Remember that normalization (also sometimes called **_Standardization_** or **_Scaling_**) means making sure that all of your data is represented at the same scale. The most common way to do this is to convert all numerical values to z-scores. 

Since KNN is a distance-based classifier, if data is in different scales, then larger scaled features have a larger impact on the distance between points.

To scale your data, use `StandardScaler` found in the `sklearn.preprocessing` module. 

In the cell below:

* Import and instantiate `StandardScaler` 
* Use the scaler's `.fit_transform()` method to create a scaled version of the training dataset  
* Use the scaler's `.transform()` method to create a scaled version of the test dataset  
* The result returned by `.fit_transform()` and `.transform()` methods will be numpy arrays, not a pandas DataFrame. Create a new pandas DataFrame out of this object called `scaled_df`. To set the column names back to their original state, set the `columns` parameter to `one_hot_df.columns` 
* Print the head of `scaled_df` to ensure everything worked correctly 

In [24]:
# Import StandardScaler
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Transform the training and test sets
scaled_data_train = scaler.fit_transform(X_train)
scaled_data_test = scaler.transform(X_test)

# Convert into a DataFrame
scaled_df_train = pd.DataFrame(scaled_data_train, columns=one_hot_df.columns)
scaled_df_train.head()

C:\Users\deon\Anaconda3\envs\learn-env\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\deon\Anaconda3\envs\learn-env\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)
C:\Users\deon\Anaconda3\envs\learn-env\lib\site-packages\ipykernel_launcher.py:9: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  if __name__ == '__main__':


Pclass       Sex       Age     SibSp     Parch      Fare  Embarked_C  \
0  0.815528 -1.390655 -0.575676 -0.474917 -0.480663 -0.500108   -0.483046   
1 -0.386113 -1.390655  1.550175 -0.474917 -0.480663 -0.435393   -0.483046   
2 -0.386113  0.719086 -0.120137 -0.474917 -0.480663 -0.644473   -0.483046   
3 -1.587755  0.719086 -0.120137 -0.474917 -0.480663 -0.115799   -0.483046   
4  0.815528 -1.390655 -1.107139  0.413551 -0.480663 -0.356656    2.070197   

   Embarked_Q  Embarked_S  
0   -0.311768    0.620174  
1   -0.311768    0.620174  
2   -0.311768    0.620174  
3   -0.311768    0.620174  
4   -0.311768   -1.612452

You may have noticed that the scaler also scaled our binary/one-hot encoded columns, too! Although it doesn't look as pretty, this has no negative effect on the model. Each 1 and 0 have been replaced with corresponding decimal values, but each binary column still only contains 2 values, meaning the overall information content of each column has not changed.

## Fit a KNN model

Now that you've preprocessed the data it's time to train a KNN classifier and validate its accuracy. 

In the cells below:

* Import `KNeighborsClassifier` from the `sklearn.neighbors` module 
* Instantiate the classifier. For now, you can just use the default parameters  
* Fit the classifier to the training data/labels
* Use the classifier to generate predictions on the test data. Store these predictions inside the variable `test_preds` 

In [25]:
# Import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

# Instantiate KNeighborsClassifier
clf = KNeighborsClassifier()

# Fit the classifier
clf.fit(scaled_data_train, y_train)

# Predict on the test set
test_preds = clf.predict(scaled_data_test)
test_preds

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 1], dtype=int64)

## Evaluate the model

Now, in the cells below, import all the necessary evaluation metrics from `sklearn.metrics` and complete the `print_metrics()` function so that it prints out **_Precision, Recall, Accuracy, and F1-Score_** when given a set of `labels` (the true values) and `preds` (the models predictions). 

Finally, use `print_metrics()` to print the evaluation metrics for the test predictions stored in `test_preds`, and the corresponding labels in `y_test`. 

In [33]:
# Your code here 
# Import the necessary functions
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


In [34]:
# Complete the function
def print_metrics(labels, preds):
    print("Precision Score: {}".format(precision_score(labels, preds)))
    print("Recall Score: {}".format(recall_score(labels, preds)))
    print("Accuracy Score: {}".format(accuracy_score(labels, preds)))
    print("F1 Score: {}".format(f1_score(labels, preds)))
    
print_metrics(y_test, test_preds)

Precision Score: 0.7058823529411765
Recall Score: 0.7317073170731707
Accuracy Score: 0.7892376681614349
F1 Score: 0.718562874251497


> Interpret each of the metrics above, and explain what they tell you about your model's capabilities. If you had to pick one score to best describe the performance of the model, which would you choose? Explain your answer.

Write your answer below this line: 


________________________________________________________________________________

Accuracy Score: 0.7892376681614349


## Improve model performance

While your overall model results should be better than random chance, they're probably mediocre at best given that you haven't tuned the model yet. For the remainder of this notebook, you'll focus on improving your model's performance. Remember that modeling is an **_iterative process_**, and developing a baseline out of the box model such as the one above is always a good start. 

First, try to find the optimal number of neighbors to use for the classifier. To do this, complete the `find_best_k()` function below to iterate over multiple values of K and find the value of K that returns the best overall performance. 

The function takes in six arguments:
* `X_train`
* `y_train`
* `X_test`
* `y_test`
* `min_k` (default is 1)
* `max_k` (default is 25)
    
> **Pseudocode Hint**:
1. Create two variables, `best_k` and `best_score`
1. Iterate through every **_odd number_** between `min_k` and `max_k + 1`. 
    1. For each iteration:
        1. Create a new `KNN` classifier, and set the `n_neighbors` parameter to the current value for k, as determined by the loop 
        1. Fit this classifier to the training data 
        1. Generate predictions for `X_test` using the fitted classifier 
        1. Calculate the **_F1-score_** for these predictions 
        1. Compare this F1-score to `best_score`. If better, update `best_score` and `best_k` 
1. Once all iterations are complete, print the best value for k and the F1-score it achieved 

In [50]:
print(X_train)

     Pclass  Sex   Age  SibSp  Parch      Fare  Embarked_C  Embarked_Q  \
376       3    0  22.0      0      0    7.2500           0           0   
458       2    0  50.0      0      0   10.5000           0           0   
732       2    1  28.0      0      0    0.0000           0           0   
507       1    1  28.0      0      0   26.5500           0           0   
830       3    0  15.0      1      0   14.4542           1           0   
2         3    0  26.0      0      0    7.9250           0           0   
356       1    0  22.0      0      1   55.0000           0           0   
502       3    0  28.0      0      0    7.6292           0           1   
260       3    1  28.0      0      0    7.7500           0           1   
623       3    1  21.0      0      0    7.8542           0           0   
159       3    1  28.0      8      2   69.5500           0           0   
517       3    1  28.0      0      0   24.1500           0           1   
828       3    1  28.0      0      0  

In [66]:
def find_best_k(X_train, y_train, X_test, y_test, min_k=1, max_k=25):
    # Your code here
    best_k = 0
    best_score = 0.0
    for k in range(min_k, max_k+1, 2):
        knn = KNeighborsClassifier(n_neighbors=k)
        knn.fit(X_train, y_train)
        preds = knn.predict(X_test)
        f1 = f1_score(y_test, preds)  # labels and preds
        print(k, f1)
        if f1 > best_score:
            best_k = k
            best_score = f1
            
    print(f"Best value for k: {best_k}")
    print(f"F1-Score: {best_score}")
        

In [40]:
find_best_k(scaled_data_train, y_train, scaled_data_test, y_test)
# Expected Output:

# Best Value for k: 17
# F1-Score: 0.7468354430379746

1 0.7078651685393258
3 0.7116564417177914
5 0.718562874251497
7 0.7228915662650603
9 0.7380952380952381
11 0.7272727272727273
13 0.7283950617283951
15 0.7204968944099378
17 0.7468354430379746
19 0.7307692307692307
21 0.7320261437908497
23 0.7402597402597403
25 0.7320261437908497
Best value for k: 17
F1-Score: 0.7468354430379746


If all went well, you'll notice that model performance has improved by 3 percent by finding an optimal value for k. For further tuning, you can use scikit-learn's built-in `GridSearch()` to perform a similar exhaustive check of hyperparameter combinations and fine tune model performance. For a full list of model parameters, see the [sklearn documentation !](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)



## (Optional) Level Up: Iterating on the data

As an optional (but recommended!) exercise, think about the decisions you made during the preprocessing steps that could have affected the overall model performance. For instance, you were asked to replace the missing age values with the column median. Could this have affected the overall performance? How might the model have fared if you had just dropped those rows, instead of using the column median? What if you reduced the data's dimensionality by ignoring some less important columns altogether?

In the cells below, revisit your preprocessing stage and see if you can improve the overall results of the classifier by doing things differently. Consider dropping certain columns, dealing with missing values differently, or using an alternative scaling function. Then see how these different preprocessing techniques affect the performance of the model. Remember that the `find_best_k()` function handles all of the fitting; use this to iterate quickly as you try different strategies for dealing with data preprocessing! 

# Steps for K-Fold Cross Validation

In [46]:
# class sklearn.model_selection.KFold(n_splits=5, shuffle=False, random_state=None)
from sklearn.model_selection import KFold
kf = KFold(n_splits=10, shuffle=False, random_state=33)

In [61]:
kf.get_n_splits(X_train)
print(kf)

KFold(n_splits=10, random_state=33, shuffle=False)


In [65]:
for train_index, test_index in kf.split(X_train):
    print("TRAIN:", train_index, "TEST:", test_index)
    print("\n")
    X_train, X_test = train_index, test_index
    y_train, y_test = train_index, test_index

TRAIN: [ 54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251
 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269
 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287
 288 289 290 291 292 293 294 295 296 297 298

## Summary

Well done! In this lab, you worked with the classic Titanic dataset and practiced fitting and tuning KNN classification models using scikit-learn! As always, this gave you another opportunity to continue practicing your data wrangling skills and model tuning skills using Pandas and scikit-learn!